qiskit/providers/ibmq/job/ibmqjob.py

```diff
@@ -203,17 +205,19 @@ def result(self, timeout: Optional[float] = None, wait: float = 5) -> Result:
             Result object
 
         Raises:
-            JobError: if the job has failed or cancelled, or if there was some
-                unexpected failure in the server.
+            IBMQJobApiError: if the job was cancelled or there was some unexpected failure
+                in the server.
+            IBMQJobFailureError: If the job failed.
         """
         # pylint: disable=arguments-differ
         # pylint: disable=access-member-before-definition,attribute-defined-outside-init
 
         if not self._wait_for_completion(timeout=timeout, wait=wait,
                                          required_status=(JobStatus.DONE,)):
-            message = 'Job was cancelled.' if self._status is JobStatus.CANCELLED \
-                else 'Job has failed. Use job.error_message() to get more details.'
-            raise JobError('Unable to retrieve job result. ' + message)
+            if self._status is JobStatus.CANCELLED:
+                raise IBMQJobApiError('Unable to retrieve job result. Job was cancelled.')
+            raise IBMQJobFailureError('Unable to retrieve job result. Job has failed. '
+                                      'Use job.error_message() to get more details.')
 
         if not self._result:  # type: ignore[has-type]
             with api_to_job_error():
```

Comment:
`api_to_job_error()` actually catches an `ApiError` and converts it to a `JobError`. It should be updated to convert to `IBMQJobApiError`.